### ISSUE TITLE
tr.resample() can take forever for some corner cases

### ISSUE THREAD

**Issue by claudiodsf** [obspy/obspy#2407]
This is related to a bug in `scipy.fftpack.rfft`: https://github.com/scipy/scipy/issues/10228

When trace length is a prime number, `rfft` takes forever.

This happened to me trying to resample a 1-hour trace at 125 Hz:
```
60*60*125 + 1 = 450001  
```
and `450001` is a prime number.

I wonder wether one could use [scipy.fftpack.next_fast_len](https://docs.scipy.org/doc/scipy/reference/generated/scipy.fftpack.next_fast_len.html) in `tr.resample` to speed up things...

**Comment by trichter** [comment]
> I wonder wether one could use scipy.fftpack.next_fast_len in tr.resample to speed up things...

I think this could be possible, but it could be error prone.

1. Append zeros to signal by using nfft=next_fast_len(N) in FFT
2. Use nfft2 = nfft // factor for IFFT, this is probably fast, too
3. Remove last (N//factor - nfft2) samples

But, it is probably easier to just use a method from interpolate.

On a side note, shifting in frequency domain is easy, too. Therefore, resample could be just another method of Trace.interpolate, couldn't it?

**Comment by d-chambers** [comment]
I have heard of several users getting tripped up on this, thanks @claudiodsf for opening the issue.

**Comment by claudiodsf** [comment]
Here is an example showing the issue:

```python
from obspy.clients.fdsn import Client
from obspy import UTCDateTime

cl = Client('RESIF')
t0 = UTCDateTime('20190501')
t1 = t0 + 60*60
tr = cl.get_waveforms('RA', 'STSF', '*', 'HNZ', t0, t1)[0]
print(tr)
# RA.STSF.00.HNZ | 2019-05-01T00:00:00.001500Z - 2019-05-01T01:00:00.001500Z | 125.0 Hz, 450001 samples

# filter to avoid aliasing
tr.detrend('linear')
tr.filter(type='lowpass', freq=5.)

# this will take forever (comment to test the interpolate method)
tr.resample(20)

# this just works
tr.interpolate(20)

print(tr)
# RA.STSF.00.HNZ | 2019-05-01T00:00:00.001500Z - 2019-05-01T01:00:00.001500Z | 20.0 Hz, 72001 samples
```

So, `tr.interpolate()` just works in my case.

**Comment by claudiodsf** [comment]
Here is a comparison of `tr.resample()` and `tr.interpolate()`:
```python
from obspy import read
import matplotlib.pyplot as plt

tr = read()[0]
tr.filter(type='lowpass', freq=5.)
tr1 = tr.copy()
tr2 = tr.copy()
tr1.resample(20)
tr2.interpolate(20)
plt.plot(tr.times(), tr.data, label='original')
plt.plot(tr1.times(), tr1.data, label='resample')
plt.plot(tr2.times(), tr2.data, label='interpolate')
plt.legend()
plt.show()
```

What is the interest of having a `tr.resample()` method?

**Comment by krischer** [comment]
Just historical reasons and maybe to also just have a frequency domain method? I think the Lanczos/sinc interpolation method is superior in almost all cases and we should probably discourage the use of the `resample()` method.

I assume we could just be clearer in the documentation? We should of course also consider padding to the next power of 2 to speed up existing codes.

**Comment by claudiodsf** [comment]
Definitively be clear in the docs. People (like me) needing to resample will go straight to the `resample()` method, ignoring that there is a better one!

**Comment by d-chambers** [comment]
>we should probably discourage the use of the resample() method

It might be a good idea to put a note/warning in the `resample` docstring that most people will actually want to use `interpolate` instead.

**Comment by calum-chamberlain** [comment]
As an aside, I had a quick play today with dropping in pyFFTW to `resample` (in a `try.. except ImportError` fallback to scipy statement).  For a day of 100.0 Hz data, + 1 sample (86400001 samples) it looks like it is faster (on my computer) to use `resample` with a pyFFTW backend than it is to interpolate (using lanczos with `a=50`).  I don't know how instructive this really is.  But we could provide pyFFTW as a possible backend?

Without multi-threading in pyFFTW my case took around 27s, with multi-threading across 12 threads it took around 10s, and interpolation took around 43s.  I didn't wait around for resampling with the scipy fftpack implementation.

**Comment by QuLogic** [comment]
NumPy has that new pocketfft implementation that may be faster as well.

**Comment by ThomasLecocq** [comment]
@calum-chamberlain @QuLogic @claudiodsf if you guys want to prepare an all-in test case we can test/reproduce easily on different FFT-based solutions, you're welcome :-)

**Comment by claudiodsf** [comment]
I think this is fixed by https://github.com/scipy/scipy/issues/2629: using SciPy 1.4.1, the bug has disappeared.

The following code reproduces the bug with SciPy <= 1.3.2:

```python
import numpy as np
from obspy import Trace
tr = Trace(np.random.random(450001))
tr.stats.sampling_rate = 125.
tr.resample(2)
```
Note that the code above works correctly even on SciPy <=  1.3.2 if the number of points is not prime.

**Comment by trichter** [comment]
> I think this is fixed by scipy/scipy#2629

Yes, thanks to pocketfft which was also adopted by Scipy in v1.4.
Then, maybe we can close this issue, when installing a recent Scipy version is enough to circumvent the problem?

**Comment by claudiodsf** [comment]
Given the fact that it’s quite a corner case, it’s probably ok to close it.
